### PR TITLE
Use MonadRec to avoid blowing the stack

### PR DESCRIPTION
**Pull request by paf31**
@garyb @ethul Could you please review?

I think we probably want to move `tailrec` to core before merging this, but I think this provides a neat solution to the recursion issue for deep call trees.

### CODE REVIEW

**Comment by paf31** [comment]
Also, this is a sort-of breaking change due to the change in signature of `goM`, but it _might_ be the case that most practical code is using `goEff` anyway.

**Comment by ethul** [comment]
I think this is great! Thanks for the PR @paf31.

I am okay with the signature change for `goM`. If we find that we need functions that work with monads that are not stack-safe perhaps we can add them in an `Unsafe` sub-module (or something appropriately named).

On that note, what are your thoughts on removing `iterM`?

File: src/Control/Monad/Free.purs [hunk 1]
```diff
@@ -102,35 +85,34 @@ resume f = case f of
...
+    Right a -> return (Right a)
 
+-- | `goEff` is `goM` specialized to the `Eff` monad.
 goEff :: forall e f a. (Functor f) => (f (Free f a) -> Eff e (Free f a)) -> Free f a -> Eff e a
-goEff fn f = runFn6 goEffImpl resume isRight unsafeLeft unsafeRight fn f
+goEff = goM
 
--- Note: can blow the stack!
-goMC :: forall f m a. (Monad m) => Natural f m -> FreeC f a -> m a
+-- | Note: can blow the stack!
```
Comment (ethul) [inline src/Control/Monad/Free.purs]: Should we remove this comment?

**Comment by paf31** [comment]
I didn't remove `iterM` because I didn't know whether there was a use case which `goM` didn't handle. If there isn't, then let's take it out.

I need to test, but I think this also fixes #22.

**Comment by paf31** [comment]
@garyb What do you think about moving `tailrec` into core? What's missing from the usual set of requirements? I think I just have to replace `pulp` with `Grunt` and add `jsvalidate`, is that right?

**Comment by garyb** [comment]
Moving `tailrec` is good by me, and yep a Grunt build is all that's "missing".

On that note though, I've become a Gulp convert over the last few months, it definitely is a better way of doing things. Going back and changing them all over seems a bit silly and arbitrary when there is nothing actually wrong with our current Gruntfiles, but I'm certainly not precious about making them all Grunt anymore - `purescript-inject` is Gulp based, and there may be a handful of others.

**Comment by paf31** [comment]
I'm not really stuck on `grunt` either. Maybe `pulp` is fine then?

**Comment by paf31** [comment]
Ok, I moved `tailrec` over. Shall I merge this and make a release then? Minor or major?

**Comment by garyb** [comment]
Major I guess, due to the type signature change. Not too many dependants at least:

- [ ] purescript-strongcheck
- [ ] purescript-argonaut
- [ ] purescript-base

**Comment by paf31** [comment]
I can always just rename it to save the hassle of a major release. My rationale was just that this is really the most general _useful_ type of `goM`, since anything else is unpredictable in terms of stack usage.

**Comment by ethul** [comment]
If you end up going with a major release, should we consider removing `iterM`. I can't think of an immediate use-case it covers that cannot be covered by `goM`. Except that `iterM` requires `Monad` and not `MonadRec`. But there may be a case I am not thinking of. However, perhaps providing only stack-safe functions here is a good way to go?

**Comment by paf31** [comment]
If nobody uses `iterM` and `goM` is general enough, I say :+1:

**Comment by paf31** [comment]
While we're breaking stuff, I'd love to change the names from `go*` to say, `runFree*` or something.

**Comment by ethul** [comment]
Good call on the `runFree*`! I'd be open to something along those lines.

**Comment by ethul** [comment]
Maybe:

```bash
go => runFree
goM => runFreeM
goMC => runFreeCM
goEff => runFreeEff
goEffC => runFreeCEff
```

is it worth adding a `runFreeC`?

**Comment by paf31** [comment]
Sorry, I just saw your message after I committed things.

I actually removed the `*Eff` methods since they're literally just aliases now, and I can't think of any reasons why we'd need them for typechecking. I can put them back if you prefer.

I can add `runFreeC` if you like.

File: src/Control/Monad/Free.purs [hunk 2]
```diff
@@ -102,36 +78,30 @@ resume f = case f of
...
+-- | See the `MonadRec` type class for more details.
+runFreeM :: forall f m a. (Functor f, MonadRec m) => (f (Free f a) -> m (Free f a)) -> Free f a -> m a
+runFreeM fn = tailRecM \f -> 
+  case resume f of
+    Left fs -> Left <$> fn fs
+    Right a -> return (Right a)
+
+-- | `runFreeMC` is the equivalent of `runFreeM` for type constructors transformed with `Coyoneda`,
+-- | hence we have no requirement that `f` be a `Functor`.
+runFreeMC :: forall f m a. (MonadRec m) => Natural f m -> FreeC f a -> m a
```
Comment (ethul) [inline src/Control/Monad/Free.purs]: I wonder if `runFreeCM` is clearer here?

**Comment by ethul** [comment]
Might be useful to add `runFreeC` if it is not too big of a deal. And makes sense removing the `*Eff` functions. This is all looking great!

**Comment by paf31** [comment]
Ok, I think this is good to go. What do you think?

**Comment by ethul** [comment]
Looks good! Thanks @paf31

**Pull request merged**